crate/server/src/core/operations/set_attribute.rs

```diff
@@ -0,0 +1,130 @@
+        .unique_identifier
+        .as_ref()
+        .ok_or(KmsError::UnsupportedPlaceholder)?
+        .as_str()
+        .context("Set Attribute: the unique identifier must be a string")?;
+
+    let mut owm: crate::database::object_with_metadata::ObjectWithMetadata =
+        retrieve_object_for_operation(
+            uid_or_tags,
+            ObjectOperationType::GetAttributes,
+            kms,
+            user,
+            params,
+        )
+        .await?;
+    trace!(
+        "Set Attribute: Retrieved object for: {:?}",
+        serde_json::to_string(&owm)
+    );
+
+    let mut attributes = owm.attributes;
+
+    match request.new_attribute {
+        Attribute::ActivationDate(activation_date) => {
+            trace!("Set Attribute: Activation Date: {:?}", activation_date);
+            attributes.activation_date = Some(activation_date);
+        }
+        Attribute::CryptographicAlgorithm(cryptographic_algorithm) => {
+            trace!(
+                "Set Attribute: Cryptographic Algorithm: {:?}",
+                cryptographic_algorithm
+            );
+            attributes.cryptographic_algorithm = Some(cryptographic_algorithm);
+        }
+        Attribute::CryptographicLength(length) => {
+            trace!("Set Attribute: Cryptographic Length: {:?}", length);
+            attributes.cryptographic_length = Some(length);
+        }
+        Attribute::CryptographicParameters(parameters) => {
+            trace!("Set Attribute: Cryptographic Parameters: {:?}", parameters);
+            attributes.cryptographic_parameters = Some(Box::new(parameters));
+        }
+        Attribute::CryptographicDomainParameters(domain_parameters) => {
+            trace!(
+                "Set Attribute: Cryptographic Domain Parameters: {:?}",
+                domain_parameters
+            );
+            attributes.cryptographic_domain_parameters = Some(domain_parameters);
+        }
+        Attribute::CryptographicUsageMask(usage_mask) => {
+            trace!("Set Attribute: Cryptographic Usage Mask: {:?}", usage_mask);
+            attributes.cryptographic_usage_mask = Some(usage_mask);
+        }
+        Attribute::Links(links) => {
+            trace!("Set Attribute: Link: {:?}", links);
+            for link in &links {
+                attributes.set_link(link.link_type, link.linked_object_identifier.clone());
+            }
+        }
+        Attribute::State(state) => {
+            trace!("Set Attribute: State: {:?}", state);
+            owm.state = state;
+        }
+        Attribute::VendorAttributes(vendor_attributes) => {
+            trace!("Set Attribute: Vendor Attributes: {:?}", vendor_attributes);
+            for vendor_attribute in vendor_attributes {
+                attributes.set_vendor_attribute(
+                    &vendor_attribute.vendor_identification,
+                    &vendor_attribute.attribute_name,
+                    vendor_attribute.attribute_value,
+                );
+            }
```

Comment:
Here we only add vendor_attributes no ? How to 'replace' an existing one ? Is it OK to check for `attribute_name` and edit the existing one or create one if new ? Same about deletion, use `attribute_name` to delete a specific one ?

Reply:
the `set_vendor_attribute` function get the `VendorAttribute` (by its name+identification) and replace it. If no `VendorAttribute` is found, just adding a new one.
```
    pub fn set_vendor_attribute(
        &mut self,
        vendor_identification: &str,
        attribute_name: &str,
        attribute_value: Vec<u8>,
    ) -> &mut Self {
        let va = self.get_vendor_attribute_mut(vendor_identification, attribute_name);
        va.attribute_value = attribute_value;
        self
    }
```